### PR TITLE
Publisher: Take screenshot

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Changelog Description
Thumbnail widget in Publisher can take a screenshot.

## Additional info
Thanks to @BigRoy we have a screenshot overlay widget.

## Todos
- [x] hide publisher window for screengrab time
- [x] add PySide6 support
- [ ] test on macOS
- [ ] test on linux

## Testing notes:
1. Open Publisher in host (e.g. traypublisher)
2. Click on take screenshot icon in thumbnail widget
3. Choose are of screengrab
4. Validate it is there

### CODE REVIEW

**Comment by BigRoy** [comment]
Thanks for setting up a draft this quickly. As mentioned on discord I'll take a look into this PR and probably refactor it a bit, because I'd prefer to have options to:

- Browse... / Upload...,
- Screenshot (which is now implemented in this PR), 
- Copy image from clipboard (e.g. with grayed out menu entry if there is none)
- Remove an already set thumbnail/screenshot, like "clear current thumbnail" (also grayed out if there is none)

A minimal example would be:

```python
has_existing_thumbnail = False  # to be implemented

menu = QtWidgets.QMenu()
action = menu.addAction("Browse...")
action = menu.addAction("Capture thumbnail")
action = menu.addAction("Paste image from clipboard")
clipboard = QtWidgets.QApplication.clipboard()
has_image_in_clipboard = not clipboard.pixmap().isNull()
action.setEnabled(has_image_in_clipboard)

menu.addSeparator()

action = menu.addAction("Clear current thumbnail")
action.setEnabled(has_existing_thumbnail)

menu.exec_()
```

image

File: openpype/tools/publisher/widgets/thumbnail_widget.py [hunk 1]
```diff
@@ -438,6 +452,15 @@ def _on_clear_clicked(self):
...
         self.thumbnail_cleared.emit()
 
+    def _on_take_screenshot(self):
+        window = self.window()
+        window.setWindowState(QtCore.Qt.WindowMinimized)
+        output_path = os.path.join(
+            self._output_dir, uuid.uuid4().hex + ".png")
+        if capture_to_file(output_path):
+            self.thumbnail_created.emit(output_path)
+        window.setWindowState(QtCore.Qt.WindowActive)
```
Comment (BigRoy) [inline openpype/tools/publisher/widgets/thumbnail_widget.py:462]:
```suggestion
        window = self.window()
        states = window.windowStates()
        window.setWindowState(QtCore.Qt.WindowMinimized)
        output_path = os.path.join(
            self._output_dir, uuid.uuid4().hex + ".png")
        if capture_to_file(output_path):
            self.thumbnail_created.emit(output_path)
        # restore original window state
        window.setWindowStates(states)
```

Would it work to just restore the original states afterwards? Just to ensure if e.g. it was maximized or whatever it'd be again?

Another approach might be to not touch the window state at all and just set window visibility (which likely inheritly sets the states too?):

```python
        window = self.window()
        visibility = window.visibility()
        window.setVisibility(QtWidgets.QWindow.Hidden)
        output_path = os.path.join(
            self._output_dir, uuid.uuid4().hex + ".png")
        if capture_to_file(output_path):
            self.thumbnail_created.emit(output_path)
        # restore original window state
        window.setVisibility(visibility)
```

**Comment by iLLiCiTiT** [comment]
> The menu would appear when e.g. left and/or right clicking on the thumbnail placeholder top right here:

If there is not a visual identication, it is non-existing feature. There must be something that tells user "you can click here to show more options".

> Browse... / Upload...

Can be added in future PR.

> Copy image from clipboard (e.g. with grayed out menu entry if there is none)

Possible. If it would be an icon it would be hard to achieve the gayout, there would have to be timer which periodically checks if clipboard has valid mime data, future PR. For menu easier...

> Remove an already set thumbnail/screenshot, like "clear current thumbnail" (also grayed out if there is none)

That's already there. The button is hidden if there is no thumbnail.

**Comment by fabiaserra** [comment]
> If there is not a visual identication, it is non-existing feature. There must be something that tells user "you can click here to show more options".

Would it help perhaps if the thumbnail area was "hoverable" and higlighted once you hover over and it's a left-click instead of right-click? On top of that we could add a click icon or three dots/lines on the thumbnail area to make it even more obvious?

**Comment by BigRoy** [comment]
> If there is not a visual identication, it is non-existing feature. There must be something that tells user "you can click here to show more options".

I'd say, change the icon on the thumbnail area to the photo camera? Maybe it also has a little text hint in it saying "Click to add thumbnail" (with maybe a "Drag 'n' Drop" support explanation too). Too me it supporting drag 'n' drop is just as 'unknown' as being able to left click on it. No? I'd actually think I'd click on it before trying a drag drop.


> Possible. If it would be an icon it would be hard to achieve the gayout, there would have to be timer which periodically checks if clipboard has valid mime data, future PR. For menu easier...

Yup, +1 for menu.

> That's already there. The button is hidden if there is no thumbnail.

Ah great - hadn't tested it in practice recently.

---

> Would it help perhaps if the thumbnail area was "hoverable" and higlighted once you hover over and it's a left-click instead of right-click? On top of that we could add a click icon or three dots/lines on the thumbnail area to make it even more obvious?

Yup, that. Plus left click I'd prefer to instead of right click. It should be clearly something you can interact with, so highlight on hover would be great.

**Comment by fabiaserra** [comment]
+1 to the "Click to add thumbnail"

**Comment by BigRoy** [comment]
Got this error in TrayPublisher:
```python
Traceback (most recent call last):
  File "E:\openpype\OpenPype\openpype\tools\publisher\widgets\thumbnail_widget.py", line 457, in _on_take_screenshot
    states = window.windowStates()
AttributeError: 'TrayPublishWindow' object has no attribute 'windowStates'
```

Same in Maya:
```python
# Traceback (most recent call last):
#   File "E:\openpype\OpenPype\openpype\tools\publisher\widgets\thumbnail_widget.py", line 457, in _on_take_screenshot
#     states = window.windowStates()
# AttributeError: 'PublisherWindow' object has no attribute 'windowStates'
```

Likely I broke it again with one my recommendations? Haha. It's clearly available on [`QWindow.windowStates`](https://doc.qt.io/qt-6/qwindow.html#windowStates) (even in [Qt5](https://doc.qt.io/qt-5/qwindow.html))?

---

Also, I just wanted to say that actually the current design I quite like. ❤️  I had expected the button just be outside of the thumbnail view but it's quite clear with the camera icon as it is. So, nice job!

image

I was thinking however to also add the _triple dots_ (for extra options) icon in there which would show the menu for the other remaining options? Since I'd still prefer it to expose the other options.

Or looking at this design @fabiaserra would you still prefer the full thumbnail view be clickable with a menu to show up

**Comment by fabiaserra** [comment]
> Or looking at this design @fabiaserra would you still prefer the full thumbnail view be clickable with a menu to show up.

Oh no, this is looking great!! thank you!!

**Comment by iLLiCiTiT** [comment]
Closed in favor of https://github.com/ynput/OpenPype/pull/5439

**Pull request closed without merging**